### PR TITLE
[Inductor] Fix the High Order Op layout issue

### PR DESCRIPTION
**Pull request by leslie-fang-intel**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #128275

Fix the issue: https://github.com/pytorch/pytorch/issues/127995

- In current implementation of creating `FallbackKernel`, the `device` of the `NoneLayout` is set to `None` when `example_output` returns from `cls.process_kernel` is `None`. https://github.com/pytorch/pytorch/blob/921aa194c77f5279b15415eaa213813ddcdb3b29/torch/_inductor/ir.py#L5632-L5649
- If a `ExternalKernel schedulerNode` has None device, the previous buffer will not flush before codegen this `ExternalKernel schedulerNode`  which causes the wrong generated code.
https://github.com/pytorch/pytorch/blob/ef2b5ed500cba0b8b2bf04e6006a0d64c910f440/torch/_inductor/scheduler.py#L2701-L2709

**Test Plan**
```
python -u -m pytest -s -v test/higher_order_ops/test_with_effects.py -k test_compile_inductor_external_op_return_none
```

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: test/higher_order_ops/test_with_effects.py [hunk 1]
```diff
@@ -198,6 +198,38 @@ def f(x):
         res = torch.compile(f, backend="inductor")(*inputs)
         self.assertTrue(torch.allclose(res, f(*inputs)))
 
+    @unittest.skipIf(IS_WINDOWS, "triton")
+    @unittest.skipIf(TEST_WITH_ROCM, "triton")
+    @unittest.skipIf(_get_torch_cuda_version() >= (11, 7), "triton")
```
Comment (eellison) [inline test/higher_order_ops/test_with_effects.py:203]: This doesn't test cuda. why are skipping here ? also, you can test directly `torch.utils._triton import has_triton`

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -5638,9 +5638,10 @@ def create(cls, kernel, *args, **kwargs):
                 unbacked_bindings,
             ) = cls.process_kernel(kernel, *args, **kwargs)
 
+        device = cls.find_device(tensor_args, example_output)
         if example_output is None:
             packed = cls(
-                NoneLayout(None),
+                NoneLayout(device),
```
Comment (eellison) [inline torch/_inductor/ir.py:5644]: Now that we're propagating device here, can we remove some of the `if device is not None` checks from the original pr:


https://github.com/pytorch/pytorch/commit/493478db4ac69179c6c8a870d573656469fcad1e#diff-d2539c9c8dc6a3d7e457767a880612e96d3c85752a77ead49a9e4e00a3e4c3c7R2443

**Review by eellison — COMMENTED** [review]
looks good, one comment

**Review by eellison — APPROVED** [review]

File: test/higher_order_ops/test_with_effects.py (hunk 1 above)
Comment (leslie-fang-intel) [reply test/higher_order_ops/test_with_effects.py:203]: I just copy these skip from UT `test_register_effectful_custom_op` in this same file. cc @zou3519 @angelayi Do you know any cornet case for `test_register_effectful_custom_op` to skip the triton testing. Remove it and let's see what是CI think about.

File: torch/_inductor/ir.py (hunk 2 above)
Comment (leslie-fang-intel) [reply torch/_inductor/ir.py:5654]: From the implementation of `FallbackKernel.find_device`, https://github.com/pytorch/pytorch/blob/685fcfb40de1bba6a99b8dc454a99ffc4b44f9a9/torch/_inductor/ir.py#L5455-L5471, I think the device may still be `None`.

File: test/higher_order_ops/test_with_effects.py (hunk 1 above)
Comment (leslie-fang-intel) [reply test/higher_order_ops/test_with_effects.py]: I think I need to skip this test on windows, the preCI reports error in `X win-vs2019-cpu-py3 / test (default, 1, 3, windows.4xlarge.nonephemeral)` as
```
2024-06-14T03:11:36.4507249Z torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
2024-06-14T03:11:36.4508166Z InvalidCxxCompiler: No working C++ compiler found in torch._inductor.config.cpp.cxx: (None, 'g++')
```

Comment (zou3519) [reply test/higher_order_ops/test_with_effects.py]: I think you just need to skip on windows

Comment (leslie-fang-intel) [reply test/higher_order_ops/test_with_effects.py]: Thanks, skip this test on Windows.

**Comment by leslie-fang-intel** [comment]
Hi @zou3519 @angelayi, looks like there is a import of this PR in https://github.com/pytorch/pytorch/commit/31df0c6c26195f3e65a6dbace0a71119063752c3, please let me know when you think it's ok to land this PR.

**Comment by angelayi** [comment]
@leslie-fang-intel Yes! It is fine to land this PR!

**Comment by zou3519** [comment]
@leslie-fang-intel thank you for the fix. I'll submit this for cherry-pick onto the 2.4 branch
